Finish migration from deprecated C-API UTType to the modern Objective-C version

brentfulgham opened the pull request:
#### 24d8cf0bd1b68d320b61fd2a1f21f48725ad4c5e
<pre>
Finish migration from deprecated C-API UTType to the modern Objective-C version
<a href="https://bugs.webkit.org/show_bug.cgi?id=293747">https://bugs.webkit.org/show_bug.cgi?id=293747</a>
<a href="https://rdar.apple.com/152250664">rdar://152250664</a>

Reviewed by Aditya Keerthi.

Mechanical change to move off of the long-deprecated C API for UTType to the
modern (supported) Objective C API.

* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::Pasteboard::readPasteboardWebContentDataForType):
(WebCore::readURLAlongsideAttachmentIfNecessary):
(WebCore::utiTypeFromCocoaType):
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
(WebCore::safeTypeForDOMToReadAndWriteForPlatformType):
(WebCore::webSafeTypes):
(WebCore::PlatformPasteboard::informationForItemAtIndex):
(WebCore::PlatformPasteboard::stringForType const):
(WebCore::PlatformPasteboard::platformPasteboardTypeForSafeTypeForDOMToReadAndWrite):
(WebCore::addRepresentationsForPlainText):
(WebCore::PlatformPasteboard::allowReadingURLAtIndex const):
(WebCore::PlatformPasteboard::write):
(WebCore::PlatformPasteboard::typesSafeForDOMToReadAndWrite const):
(WebCore::createItemProviderRegistrationList):
(WebCore::PlatformPasteboard::readString const):
(WebCore::PlatformPasteboard::readURL const):
(WebCore::PlatformPasteboard::containsURLStringSuitableForLoading):
* Source/WebCore/platform/ios/WebItemProviderPasteboard.mm:
(typeConformsToTypes):
(-[NSItemProvider web_containsFileURLAndFileUploadContent]):
(-[NSItemProvider web_fileUploadContentTypes]):
(-[WebItemProviderPasteboard _preLoadedDataConformingToType:forItemProviderAtIndex:]):
(classForTypeIdentifier):
(linkTemporaryItemProviderFilesToDropStagingDirectory):
(-[WebItemProviderPasteboard typeIdentifiersToLoad:]):
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::write):
(WebCore::Pasteboard::read):
(WebCore::Pasteboard::writeString):
(WebCore::Pasteboard::bufferConvertedToPasteboardType):
* Source/WebCore/platform/mac/PasteboardWriter.mm:
(WebCore::toUTIUnlessAlreadyUTI):
(WebCore::createPasteboardWriter):
* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::isFilePasteboardType):
(WebCore::canWritePasteboardType):
(WebCore::PlatformPasteboard::bufferForType const):
(WebCore::urlStringsFromPasteboard):
(WebCore::typeIdentifierForPasteboardType):
(WebCore::PlatformPasteboard::allStringsForType const):
(WebCore::safeTypeForDOMToReadAndWriteForPlatformType):
(WebCore::PlatformPasteboard::setStringForType):
* Source/WebKit/Shared/Cocoa/WebIconUtilities.mm:
(WebKit::iconForFiles):
* Source/WebKit/Shared/mac/PasteboardTypes.mm:
(WebKit::PasteboardTypes::forEditing):
(WebKit::PasteboardTypes::forSelection):
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
(API::isDeclaredOrDynamicTypeIdentifier):
(API::Attachment::mimeType const):
(API::Attachment::utiType const):
(API::Attachment::setFileWrapperAndUpdateContentType):
* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView actionSheetAssistant:performAction:]):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
* Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
(-[WKSharingServicePickerDelegate sharingService:didShareItems:]):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::setupServicesMenu):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setPromisedDataForImage):

Canonical link: <a href="https://commits.webkit.org/300131@main">https://commits.webkit.org/300131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/795b932c74011ca21c6eea385727b3d430871f0e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/121408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/41105 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/31764 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/127849 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/73491 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4fd415cf-936c-42ef-b20e-e9508c634241) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/123284 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/41807 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/49684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/92220 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/61365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/124360 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/33379 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/108776 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/72907 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/327971f0-21e6-4d20-bd81-c1a112b766ac) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/32396 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/26925 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/71429 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/102875 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/27099 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/130683 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/48336 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/36756 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/100814 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/48704 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/104984 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/100719 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25535 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/46128 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/24200 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/45032 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/48194 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/53907 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/47666 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/51012 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/49348 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/PAL/pal/spi/cocoa/UniformTypeIdentifiersSPI.h:
```diff
@@ -0,0 +1,40 @@
+#include <UniformTypeIdentifiers/UTTypePriv.h>
+#else
+
+@interface UTType ()
```
Nit, since you have an empty line before `@end`.
```suggestion
@interface UTType ()

```

pxlcoder commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -59,19 +61,15 @@
-        for (NSString *uti in (__bridge NSArray<NSString *> *)allUTIs.get()) {
-            auto type = adoptCF(UTTypeCopyPreferredTagWithClass((__bridge CFStringRef)uti, kUTTagClassMIMEType));
+        [UTType _enumerateAllDeclaredTypesUsingBlock:^(UTType *utType, BOOL *) {
+            auto type = utType.preferredMIMEType;
```
Should use `RetainPtr`.

pxlcoder commented on Source/WebCore/platform/cocoa/MIMETypeRegistryCocoa.mm:
```diff
@@ -59,19 +61,15 @@
-        }
-ALLOW_DEPRECATED_DECLARATIONS_END
+                return;
+            auto extensions = (NSArray<NSString *> *)utType.tags[UTTagClassFilenameExtension];
```
This should use `dynamic_objc_cast` (or `checked_objc_cast`) and `RetainPtr`.

pxlcoder commented on Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
```diff
@@ -141,23 +141,39 @@ static bool shouldTreatAtLeastOneTypeAsFile(NSArray<NSString *> *platformTypes)
 #if PASTEBOARD_SUPPORTS_ITEM_PROVIDERS
 
-static const char *safeTypeForDOMToReadAndWriteForPlatformType(const String& platformType, PlatformPasteboard::IncludeImageTypes includeImageTypes)
+static bool platformTypeConformsToWebArchivePBoardType(NSString *platformType, const RetainPtr<UTType>& platformUTType)
```
Not too common to pass in a `RetainPtr` parameter, any reason to not make the parameter `UTType * platformUTType`?

brentfulgham replied:
No, I will correct that.

pxlcoder commented on Source/WebCore/platform/ios/WebItemProviderPasteboard.mm:
```diff
@@ -53,13 +53,32 @@
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-        if (UTTypeConformsTo((__bridge CFStringRef)type, (__bridge CFStringRef)conformsToType))
+        // LS recognizes identical type strings as conforming
+        if ([type isEqualToString:conformsToType])
```
Is this just an optimization to avoid creating `conformsToUTType` below?

brentfulgham replied:
No -- it's actually a difference between the old C SPI and the Objective-C API. `UTTypeConformsTo` always did an early return (`YES`) if the passed `CFStringRef`s were the same, even if the type was not a true registered UTType. Several test cases revealed that we depend on this behavior.

The Objective-C API only returns `YES` if the types both resolve to registered `UTType` objects, and those types conform.

pxlcoder replied:
Interesting! I think it's worth mentioning that in the commit message.

brentfulgham replied:
I added this as a comment in the code, for future readers.

pxlcoder commented on Source/WebCore/platform/ios/WebItemProviderPasteboard.mm:
```diff
@@ -741,7 +755,7 @@ - (NSInteger)numberOfFiles
     auto urlExtension = url.pathExtension;
     if (!urlExtension.length)
-        urlExtension = [UTType typeWithIdentifier:typeIdentifier].preferredFilenameExtension;
+        urlExtension = utType.get().preferredFilenameExtension;
```
```suggestion
        urlExtension = [utType preferredFilenameExtension];
```

pxlcoder commented on Source/WebCore/platform/mac/PasteboardWriter.mm:
```diff
@@ -46,12 +46,11 @@
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    if (UTTypeIsDeclared(bridge_cast(type)) || UTTypeIsDynamic(bridge_cast(type))) {
+    RetainPtr utType = [UTType typeWithIdentifier:type];
+    if (utType.get().declared || utType.get().dynamic) {
```
```suggestion
    if ([utType declared] || [utType dynamic]) {
```

brentfulgham replied:
@pxlcoder Amazing -- I could NOT get this to build until I reread the documentation, and realized the property's accessor is actually `isDeclared` and `isDynamic`. Crazy that the pointer-dereference access through `get()` figured it out, but trying to use the Objc syntax wouldn't work!

pxlcoder commented on Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
```diff
@@ -50,22 +51,19 @@ static bool isFilePasteboardType(const String& type)
-    auto cfString = type.createCFString();
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    if (UTTypeIsDeclared(cfString.get()) || UTTypeIsDynamic(cfString.get()))
+    auto nsString = type.createNSString();
```
Explicit smart pointers preferred over `auto`.
```suggestion
    RetainPtr nsString = type.createNSString();
```

pxlcoder commented on Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
```diff
@@ -50,22 +51,19 @@ static bool isFilePasteboardType(const String& type)
-    if (UTTypeIsDeclared(cfString.get()) || UTTypeIsDynamic(cfString.get()))
+    auto nsString = type.createNSString();
+    RetainPtr utType = [UTType typeWithIdentifier:nsString.get()];
+    if (utType.get().declared || utType.get().dynamic)
```
```suggestion
    if ([utType declared] || [utType dynamic])
```

pxlcoder commented on Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
```diff
@@ -216,30 +210,27 @@ static bool pasteboardMayContainFilePaths(NSPasteboard *pasteboard)
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    if (UTTypeIsDeclared(pasteboardType.createCFString().get()))
+    RetainPtr utType = [UTType typeWithIdentifier:pasteboardType.createNSString().get()];
+    if (utType.get().declared)
```
```suggestion
    if ([utType declared])
```

pxlcoder commented on Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
```diff
@@ -255,19 +246,21 @@ static String typeIdentifierForPasteboardType(const String& pasteboardType)
+        || [platformType  isEqualToString:legacyRTFDPasteboardType()] || [platformType isEqualToString:legacyRTFPasteboardType()])
+        return "text/html"_s;
+
+    auto nsWebArchivePboardType = String(WebArchivePboardType).createNSString();
```
```suggestion
    RetainPtr nsWebArchivePboardType = String(WebArchivePboardType).createNSString();
```

pxlcoder commented on Source/WebCore/platform/network/mac/UTIUtilities.h:
```diff
@@ -27,11 +27,18 @@
 
+#ifdef __OBJC__
+@class UTType;
+#endif
```
```suggestion
OBJC_CLASS UTType;
```

pxlcoder commented on Source/WebCore/platform/network/mac/UTIUtilities.mm:
```diff
@@ -63,39 +65,42 @@ String MIMETypeFromUTI(const String& uti)
+ALLOW_DEPRECATED_DECLARATIONS_BEGIN
+    // Note: We have to silence deprecated declarations because some types this
+    // method returns are deprecated.
+    RetainPtr<NSOrderedSet<UTType *>> parenttypes = utType._parentTypes;
```
```suggestion
    RetainPtr<NSOrderedSet<UTType *>> parentTypes = utType._parentTypes;
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
```diff
@@ -74,25 +73,28 @@ static BOOL isDeclaredOrDynamicTypeIdentifier(NSString *type)
-ALLOW_DEPRECATED_DECLARATIONS_END
+    if (RetainPtr utType = [UTType typeWithIdentifier:contentType.get()]) {
+        if (isDeclaredOrDynamicTypeIdentifier(utType.get()))
+            return utType.get().preferredMIMEType;
```
```suggestion
            return [utType preferredMIMEType];
```

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
```diff
@@ -74,25 +73,28 @@ static BOOL isDeclaredOrDynamicTypeIdentifier(NSString *type)
+    }
+
+    RetainPtr preferredType = [UTType typeWithMIMEType:contentType.get()];
+    return preferredType.get().identifier;
```
```suggestion
    return [preferredType identifier];
```

brentfulgham committed:
Finish migration from deprecated C-API UTType to the modern Objective-C version
https://bugs.webkit.org/show_bug.cgi?id=293747
rdar://152250664

Reviewed by Aditya Keerthi.

Mechanical change to move off of the long-deprecated C API for UTType to the
modern (supported) Objective C API.

* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::Pasteboard::readPasteboardWebContentDataForType):
(WebCore::readURLAlongsideAttachmentIfNecessary):
(WebCore::utiTypeFromCocoaType):
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
(WebCore::safeTypeForDOMToReadAndWriteForPlatformType):
(WebCore::webSafeTypes):
(WebCore::PlatformPasteboard::informationForItemAtIndex):
(WebCore::PlatformPasteboard::stringForType const):
(WebCore::PlatformPasteboard::platformPasteboardTypeForSafeTypeForDOMToReadAndWrite):
(WebCore::addRepresentationsForPlainText):
(WebCore::PlatformPasteboard::allowReadingURLAtIndex const):
(WebCore::PlatformPasteboard::write):
(WebCore::PlatformPasteboard::typesSafeForDOMToReadAndWrite const):
(WebCore::createItemProviderRegistrationList):
(WebCore::PlatformPasteboard::readString const):
(WebCore::PlatformPasteboard::readURL const):
(WebCore::PlatformPasteboard::containsURLStringSuitableForLoading):
* Source/WebCore/platform/ios/WebItemProviderPasteboard.mm:
(typeConformsToTypes):
(-[NSItemProvider web_containsFileURLAndFileUploadContent]):
(-[NSItemProvider web_fileUploadContentTypes]):
(-[WebItemProviderPasteboard _preLoadedDataConformingToType:forItemProviderAtIndex:]):
(classForTypeIdentifier):
(linkTemporaryItemProviderFilesToDropStagingDirectory):
(-[WebItemProviderPasteboard typeIdentifiersToLoad:]):
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::write):
(WebCore::Pasteboard::read):
(WebCore::Pasteboard::writeString):
(WebCore::Pasteboard::bufferConvertedToPasteboardType):
* Source/WebCore/platform/mac/PasteboardWriter.mm:
(WebCore::toUTIUnlessAlreadyUTI):
(WebCore::createPasteboardWriter):
* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::isFilePasteboardType):
(WebCore::canWritePasteboardType):
(WebCore::PlatformPasteboard::bufferForType const):
(WebCore::urlStringsFromPasteboard):
(WebCore::typeIdentifierForPasteboardType):
(WebCore::PlatformPasteboard::allStringsForType const):
(WebCore::safeTypeForDOMToReadAndWriteForPlatformType):
(WebCore::PlatformPasteboard::setStringForType):
* Source/WebKit/Shared/Cocoa/WebIconUtilities.mm:
(WebKit::iconForFiles):
* Source/WebKit/Shared/mac/PasteboardTypes.mm:
(WebKit::PasteboardTypes::forEditing):
(WebKit::PasteboardTypes::forSelection):
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
(API::isDeclaredOrDynamicTypeIdentifier):
(API::Attachment::mimeType const):
(API::Attachment::utiType const):
(API::Attachment::setFileWrapperAndUpdateContentType):
* Source/WebKit/UIProcess/ios/WKPDFView.mm:
(-[WKPDFView actionSheetAssistant:performAction:]):
* Source/WebKit/UIProcess/ios/forms/WKFileUploadPanel.mm:
* Source/WebKit/UIProcess/mac/WKSharingServicePickerDelegate.mm:
(-[WKSharingServicePickerDelegate sharingService:didShareItems:]):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::setupServicesMenu):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::WebViewImpl::setPromisedDataForImage):

Canonical link: https://commits.webkit.org/300131@main
